packages/pros/src/adi/mod.rs

```diff
@@ -1,56 +1,68 @@
-use core::{
-    ffi::c_int,
-    ops::{Deref, DerefMut},
-};
-
-pub struct AdiPort(u8);
-
-impl AdiPort {
-    /// Create an AdiPort without checking if it is valid.
-    ///
-    /// # Safety
-    ///
-    /// The port must be above 0 and below [`pros_sys::NUM_ADI_PORTS`].
-    pub unsafe fn new_unchecked(port: u8) -> Self {
-        Self(port)
-    }
-    /// Create an AdiPort, returning `None` if the port is invalid.
-    pub fn try_new(port: u8) -> Option<Self> {
-        if c_int::from(port) < pros_sys::NUM_ADI_PORTS {
-            Some(Self(port))
-        } else {
-            None
-        }
-    }
-    /// Create an AdiPort.
-    ///
-    /// # Panics
-    ///
-    /// Panics if the port is greater than or equal to [`pros_sys::NUM_ADI_PORTS`].
-    pub fn new(port: u8) -> Self {
-        Self::try_new(port).expect("Invalid ADI port")
-    }
-}
+//! ADI (TriPort) devices on the Vex V5.
+//! 
+//! Most ADI devices can be created with a `new` function that generally takes a port number.
+//! Devi
+//! 
 
-impl Deref for AdiPort {
-    type Target = u8;
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
+use crate::error::{map_errno, PortError};
+
+
+use snafu::Snafu;
 
-impl DerefMut for AdiPort {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
-    }
-}
 
-pub struct AdiAnalogIn {
-    port: AdiPort,
+
+pub mod port;
+pub mod analog;
+pub mod digital;
+
+pub mod encoder;
+pub mod motor;
+pub mod ultrasonic;
+pub mod gyro;
+pub mod potentiometer;
+
+
+#[derive(Debug, Snafu)]
+pub enum AdiError {
+    #[snafu(display("Another resource is currently trying to access the ADI."))]
+    AlreadyInUse,
+
+    #[snafu(display("The port specified has been reconfigured or is not configured for digital input."))]
+    DigitalInputNotConfigured,
+
+    #[snafu(display("The port specified cannot be configured due to an invalid configuration type."))]
+    InvalidConfigType,
+
+    #[snafu(display("The port has already been configured."))]
+    AlreadyConfigured,
+
+    #[snafu(display("The port specified is invalid."))]
+    InvalidPort,
+
+    #[snafu(display("Generic exception."))]
+    Generic,
```

Comment:
This preferably shouldn't exist because it tells the user nothing about the error that happened.